packages/tools/src/lib/build-cache/common.ts

```diff
@@ -10,28 +10,79 @@ export type SupportedRemoteCacheProviders = 'github-actions';
 
 export type RemoteArtifact = {
   name: string;
-  downloadUrl: string;
+  url: string;
+  id?: string;
 };
 
 export type LocalArtifact = {
   name: string;
   path: string;
 };
 
+/**
+ * Interface for implementing remote build cache providers.
+ * Remote cache providers allow storing and retrieving native build artifacts (e.g. APK, IPA)
+ * from remote storage like S3, GitHub Artifacts etc.
+ */
 export interface RemoteBuildCache {
+  /** Unique identifier for this cache provider, will be displayed in logs */
   name: string;
-  query({
+
+  /**
+   * List available artifacts matching the given name pattern
+   * @param artifactName - Passed after fingerprinting the build, e.g. `rnef-android-debug-1234567890` for android in debug variant
+   * @param limit - Optional maximum number of artifacts to return
+   * @returns Array of matching remote artifacts, or null if none found
+   */
+  list({
     artifactName,
+    limit,
   }: {
-    artifactName: string;
-  }): Promise<RemoteArtifact | null>;
+    artifactName: string | undefined;
+    limit?: number;
+  }): Promise<RemoteArtifact[] | null>;
+
+  /**
+   * Download a remote artifact to local storage
+   * @param artifact - Remote artifact to download, as returned by `list` method
+   * @param loader - Optional progress indicator
+   * @returns Local artifact info after download
+   */
   download({
     artifact,
     loader,
   }: {
     artifact: RemoteArtifact;
-    loader: ReturnType<typeof spinner>;
+    loader?: ReturnType<typeof spinner>;
   }): Promise<LocalArtifact>;
+
+  /**
+   * Delete a remote artifact
+   * @param artifact - Remote artifact to delete, as returned by `list` method
+   * @param loader - Optional progress indicator
+   * @returns True if deletion was successful
+   */
+  delete({
+    artifact,
+    loader,
+  }: {
+    artifact: RemoteArtifact;
+    loader?: ReturnType<typeof spinner>;
+  }): Promise<boolean>;
+
+  /**
+   * Upload a local artifact to remote storage
+   * @param artifact - Local artifact to upload, as returned by `download` method
+   * @param loader - Optional progress indicator
+   * @returns Remote artifact info if upload successful, null otherwise
```

Comment:
"otherwise" implies error or a different case (e.g. no support)?

maybe better to throw error instead or return an object with additional info that the CLI could display?

Reply:
The upload is something I haven't given too much thought yet, as GitHub API doesn't support it. I'll make it return `RemoteArtifact` or throw